libinithooks/inithooks_log.py

```diff
@@ -0,0 +1,36 @@
+import os
+import subprocess
+from dataclasses import dataclass
+
+INITHOOK_LOG = os.getenv("INITHOOKS_LOGFILE", "/var/log/inithooks.log")
+LOG_LEVELS = ["err", "warn", "info", "debug"]
+
+
+class InitLogError(Exception):
+    pass
+
+
+@dataclass
+class InitLog:
+    inithook_name: str
+    log_file: str = INITHOOK_LOG
+
+    def write(self, msg: str, level: str = "info") -> None:
+        """Write to log & journal
+        valid levels: err|warn|info|debug
+        """
+        if level not in LOG_LEVELS:
+            raise InitLogError(f"invalid log level '{level}'")
+        msg = f"[{self.inithook_name}] {msg}".rstrip()
+        subprocess.run(
+            [
+                "/usr/bin/logger",
+                "-t",
+                "inithooks",
+                "-p",
+                level,
+                msg,
+            ]
+        )
+        with open(self.log_file, "a") as fob:
+            fob.write(f"{msg}\n")
```

Comment:
I think this is superfluous as we're already using the stdlib logging library in dialog. I'm happy to leave this as is for now, to not block the nextcloud PR, as this is not really an issue issue.

However it might be worth having a look at the python howto logging doc page: https://docs.python.org/3/howto/logging.html

Basically the logging allows global config with the ``logging.basicConfig`` function.
You can get loggers for different purposes, they form some kind of tree of loggers.

```python3
import logging

root = logging.getLogger() # root logger
foo = logging.getLogger("foo") # child of root logger, called "foo"
bar = logging.getLogger("foo.bar") # child of "foo" logger called "bar"

assert(root.getChild("foo") == foo)
assert(foo.getChild("bar") == bar)
assert(bar.parent == foo)
assert(foo.parent == root)

# configuration flows from parent to child so
root.setLevel(logging.DEBUG)
assert(root.level == logging.DEBUG)
assert(foo.level == logging.DEBUG)
assert(bar.level == logging.DEBUG)

# but not upwards
foo.setLevel(logging.WARNING)
assert(root.level == logging.DEBUG)
assert(foo.level == logging.WARNING)
assert(bar.level == logging.WARNING)

# logging is pretty simple
foo.debug("...")
foo.info("...")
foo.warning("...")
foo.error("...")
foo.fatal("...")

# changing the format of the logs is likewise pretty simple
logging.basicConfig(format='%(asctime)s %(name) %(levelname)s: %(message)s', level=logging.DEBUG)
# basicConfig won't change stuff after initial config unless you set `force=True`
bar.warn("test")
# outputs "2024-07-17 17:07:00,684 foo.bar WARNING: test"

# all logs are written to all handlers configured for a given logger
root.addHandler(logging.FileHandler(filename='/var/log/test'))
root.addHandler(logging.handlers.SysLogHandler())
# etc.

Reply:
Yep, good call. FWIW I've opened an issue on the tracker regarding this: https://github.com/turnkeylinux/tracker/issues/1973